File: python/baize/moves.py
"""Legal move generator.

Ports the Rust move generation from engine/src/moves.rs:
  - legal_moves(session) -> list[LegalMove]
  - Grid-based movement: step, slide, leap, hop
  - Direction resolution, friendly blocking, enemy capture
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from baize.action import Action
from baize.definition import AdjacencyLiteral, Component, DirectionNameLiteral, MovementPrimitive
from baize.runtime import (
    ComponentId,
    GameSession,
    GridZone,
    RuntimeZone,
)

MAX_LEGAL_MOVES = 10_000
"""Maximum number of legal moves to generate before stopping."""


@dataclass
class LegalMove:
    """A legal move that a player can make."""

    component_id: ComponentId
    action: Action


def legal_moves(session: GameSession) -> list[LegalMove]:
    """Generate all legal moves for the current player.

    Returns at most MAX_LEGAL_MOVES moves to prevent combinatorial explosion.
    """
    assert isinstance(session, GameSession), (
        f"session must be GameSession, got {type(session).__name__}"
    )
    player = session.current_player()
    if player is None:
        return []

    moves: list[LegalMove] = []

    # For each component owned by the current player on the board, generate moves
    for zone_name, zone in session.runtime.zones.items():
        if len(moves) >= MAX_LEGAL_MOVES:
            break
        if zone_name not in session.definition.zones:
            continue

        if isinstance(zone, GridZone):
            seen: set[ComponentId] = set()
            for row in range(zone.height):
                if len(moves) >= MAX_LEGAL_MOVES:
                    break
                for col in range(zone.width):
                    if len(moves) >= MAX_LEGAL_MOVES:
                        break
                    idx = row * zone.width + col
                    cid = zone.cells[idx]
                    if cid is None:
                        continue
                    if cid in seen:
                        continue  # Skip duplicate cells of spanning component
                    seen.add(cid)
                    comp_data = session.runtime.components.get(cid)
                    if comp_data is None:
                        continue
                    if comp_data.owner != player:
                        continue
                    comp_def = session.definition.components.get(
                        comp_data.component_type
                    )
                    if comp_def is None:
                        continue
                    zone_def = session.definition.zones[zone_name]
                    _generate_grid_moves(
                        session,
                        zone_name,
                        cid,
                        comp_def,
                        col,
                        row,
                        zone_def.adjacency,
                        moves,
                    )

    # Also check per-player zones
    if len(moves) < MAX_LEGAL_MOVES:
        player_state = session.runtime.players.get(player)
        if player_state is not None:
            for _zone_name, _zone in player_state.zones.items():
                if len(moves) >= MAX_LEGAL_MOVES:
                    break
                _generate_hand_plays(session, player, _zone_name, _zone, moves)

    return moves


def _generate_grid_moves(
    session: GameSession,
    zone_name: str,
    cid: ComponentId,
    comp_def: Component,
    col: int,
    row: int,
    adjacency: AdjacencyLiteral | None,
    moves: list[LegalMove],
) -> None:
    """Generate moves for a piece on a grid using its movement primitives."""
    zone = session.runtime.zones.get(zone_name)
    if zone is None or not isinstance(zone, GridZone):
        return

    comp_data = session.runtime.components.get(cid)
    player = comp_data.owner if comp_data is not None else ""
    if player is None:
        player = ""

    for mp in comp_def.movement:
        if len(moves) >= MAX_LEGAL_MOVES:
            return
        if mp.primitive == "step":
            dirs = _resolve_directions(mp, player, adjacency)
            dist = mp.distance if mp.distance is not None else 1
            for dx, dy in dirs:
                nx = col + dx * dist
                ny = row + dy * dist
                if not zone._cell_valid(nx, ny):
                    continue
                if _check_cell_condition(zone, session, nx, ny, mp, player):
                    moves.append(
                        LegalMove(
                            component_id=cid,
                            action=_make_move_action(zone_name, col, row, nx, ny),
                        )
                    )

        elif mp.primitive == "slide":
            dirs = _resolve_directions(mp, player, adjacency)
            for dx, dy in dirs:
                nx = col + dx
                ny = row + dy
                while zone._cell_valid(nx, ny) and len(moves) < MAX_LEGAL_MOVES:
                    target = zone.grid_get(nx, ny)
                    if target is not None:
                        if _is_enemy(session, target, player):
                            moves.append(
                                LegalMove(
                                    component_id=cid,
                                    action=_make_move_action(
                                        zone_name, col, row, nx, ny
                                    ),
                                )
                            )
                        break
                    moves.append(
                        LegalMove(
                            component_id=cid,
                            action=_make_move_action(zone_name, col, row, nx, ny),
                        )
                    )
                    nx += dx
                    ny += dy

        elif mp.primitive == "leap":
            ldx = mp.dx
            ldy = mp.dy
            if ldx is not None and ldy is not None:
                for sx, sy in [(1, 1), (1, -1), (-1, 1), (-1, -1)]:
                    for adx, ady in [(ldx, ldy), (ldy, ldx)]:
                        nx = col + adx * sx
                        ny = row + ady * sy
                        if not zone._cell_valid(nx, ny):
                            continue
                        target = zone.grid_get(nx, ny)
                        if target is None:
                            moves.append(
                                LegalMove(
                                    component_id=cid,
                                    action=_make_move_action(
                                        zone_name, col, row, nx, ny
                                    ),
                                )
                            )
                        elif _is_enemy(session, target, player):
                            moves.append(
                                LegalMove(
                                    component_id=cid,
                                    action=_make_move_action(
                                        zone_name, col, row, nx, ny
                                    ),
                                )
                            )

        elif mp.primitive == "hop":
            dirs = _resolve_directions(mp, player, adjacency)
            for dx, dy in dirs:
                mid_x = col + dx
                mid_y = row + dy
                land_x = col + dx * 2
                land_y = row + dy * 2
                if not zone._cell_valid(mid_x, mid_y):
                    continue
                if not zone._cell_valid(land_x, land_y):
                    continue
                # Middle cell must be occupied
                if zone.grid_get(mid_x, mid_y) is None:
                    continue
                # Landing cell must be empty
                if zone.grid_get(land_x, land_y) is None:
                    moves.append(
                        LegalMove(
                            component_id=cid,
                            action=_make_move_action(
                                zone_name, col, row, land_x, land_y
                            ),
                        )
                    )


def _generate_hand_plays(
    session: GameSession,
    player: str,
    zone_name: str,
    zone: RuntimeZone,
    moves: list[LegalMove],
) -> None:
    """Generate moves for playing from a hand zone (stub)."""
    # Hand plays are game-specific; will be expanded


def _resolve_directions(
    mp: MovementPrimitive,
    player: str,
    adjacency: AdjacencyLiteral | None = None,
) -> list[tuple[int, int]]:
    """Resolve direction names to (dx, dy) vectors, respecting zone adjacency."""
    direction = mp.direction
    if direction is None:
        return _adjacent_directions(adjacency)

    if isinstance(direction, str):
        return _direction_vectors(direction, adjacency)
    if isinstance(direction, list):
        result: list[tuple[int, int]] = []
        for name in direction:
            result.extend(_direction_vectors(name, adjacency))
        return result
    return _adjacent_directions(adjacency)


def _direction_vectors(
    name: str, adjacency: AdjacencyLiteral | None = None
) -> list[tuple[int, int]]:
    """Convert a direction name to (dx, dy) vectors."""
    if name == "orthogonal":
        return [(1, 0), (-1, 0), (0, 1), (0, -1)]
    if name == "diagonal":
        if adjacency == "hex_6":
            return [(-1, 1), (1, -1)]
        return [(1, 1), (1, -1), (-1, 1), (-1, -1)]
    if name == "adjacent":
        return _adjacent_directions(adjacency)
    if name == "forward":
        return [(0, 1)]
    if name == "forward_diagonal":
        return [(1, 1), (-1, 1)]
    if name == "backward":
        return [(0, -1)]
    if name == "backward_diagonal":
        return [(1, -1), (-1, -1)]
    return _adjacent_directions(adjacency)


_HEX_DIRECTIONS: list[tuple[int, int]] = [
    (-1, 0), (1, 0), (0, -1), (0, 1), (-1, 1), (1, -1)
]


def _adjacent_directions(
    adjacency: AdjacencyLiteral | None = None,
) -> list[tuple[int, int]]:
    """Return neighbor directions based on adjacency model."""
    if adjacency == "hex_6":
        return list(_HEX_DIRECTIONS)
    if adjacency == "orthogonal_4":
        return [(1, 0), (-1, 0), (0, 1), (0, -1)]
    return [
        (1, 0),
        (-1, 0),
        (0, 1),
        (0, -1),
        (1, 1),
        (1, -1),
        (-1, 1),
        (-1, -1),
    ]


def _is_enemy(session: GameSession, target_id: ComponentId, player: str) -> bool:
    """Check if a component belongs to an enemy player."""
    comp = session.runtime.components.get(target_id)
    if comp is None:
        return False
    if comp.owner is None:
        return False
    return comp.owner != player


def _check_cell_condition(
    zone: GridZone,
    session: GameSession,
    col: int,
    row: int,
    mp: MovementPrimitive,
    player: str,
) -> bool:
    """Check whether a step move can land on the given cell."""
    occupant = zone.grid_get(col, row)
    condition = mp.condition if mp.condition is not None else "empty_or_enemy"

    if condition == "empty":
        return occupant is None
    if condition == "enemy":
        return occupant is not None and _is_enemy(session, occupant, player)
    if condition == "empty_or_enemy":
        return occupant is None or _is_enemy(session, occupant, player)
    # Try CEL evaluation for complex conditions
    from baize.cel import try_eval_move_condition

    cell_empty = occupant is None
    cell_enemy = occupant is not None and _is_enemy(session, occupant, player)
    cel_result = try_eval_move_condition(cell_empty, cell_enemy, condition)
    if cel_result is not None:
        return cel_result
    # Legacy fallback: allow if empty or enemy
    return occupant is None or _is_enemy(session, occupant, player)


def _make_move_action(
    zone: str, from_col: int, from_row: int, to_col: int, to_row: int
) -> Action:
    """Create a MovePiece action for a grid move."""
    return Action(
        action_type="move_piece",
        from_pos={
            "zone": zone,
            "cell": f"{from_col},{from_row}",
        },
        to_pos={
            "zone": zone,
            "cell": f"{to_col},{to_row}",
        },
    )

File: engine/src/moves.rs
use crate::action::{Action, ActionType, Position};
use crate::definition::{Adjacency, Component, DirectionName, MovementPrimitive, PrimitiveType};
use crate::runtime::{ComponentId, GameSession, RuntimeZone};

/// Maximum number of legal moves to generate before stopping.
/// Prevents combinatorial explosion on huge boards with many pieces.
const MAX_LEGAL_MOVES: usize = 10_000;

/// A legal move that a player can make.
#[derive(Debug, Clone)]
pub struct LegalMove {
    pub component_id: ComponentId,
    pub action: Action,
}

/// Generate all legal moves for the current player.
pub fn legal_moves(session: &GameSession) -> Vec<LegalMove> {
    let player = match session.current_player() {
        Some(p) => p.to_string(),
        None => return Vec::new(),
    };

    let mut moves = Vec::new();

    // For each component owned by the current player on the board, generate moves
    for (zone_name, zone) in &session.runtime.zones {
        if moves.len() >= MAX_LEGAL_MOVES {
            break;
        }
        let zone_def = match session.definition.zones.get(zone_name) {
            Some(z) => z,
            None => continue,
        };

        match zone {
            RuntimeZone::Grid { storage, .. } => {
                // Track seen components to avoid duplicate processing of multi-cell spans
                let mut seen = std::collections::HashSet::new();
                let occupied = storage.occupied_cells();
                for (col, row, cid) in occupied {
                    if moves.len() >= MAX_LEGAL_MOVES {
                        break;
                    }
                    if col < 0 || row < 0 {
                        continue; // skip negative coords for move generation
                    }
                    if !seen.insert(cid) {
                        continue; // Already processed this spanning component
                    }
                    if let Some(comp_data) = session.runtime.components.get(cid) {
                        if comp_data.owner.as_deref() != Some(&player) {
                            continue;
                        }
                        if let Some(comp_def) =
                            session.definition.components.get(&comp_data.component_type)
                        {
                            generate_grid_moves(
                                session,
                                zone_name,
                                cid,
                                comp_def,
                                col as u32,
                                row as u32,
                                zone_def.adjacency,
                                &mut moves,
                            );
                        }
                    }
                }
            }
            _ => {
                // Non-grid zones: generate placement moves for components in supply
                generate_placement_moves(session, &player, zone_name, zone_def, zone, &mut moves);
            }
        }
    }

    // Also check per-player zones
    if moves.len() < MAX_LEGAL_MOVES {
        if let Some(player_state) = session.runtime.players.get(&player) {
            for (zone_name, zone) in &player_state.zones {
                if moves.len() >= MAX_LEGAL_MOVES {
                    break;
                }
                generate_hand_plays(session, &player, zone_name, zone, &mut moves);
            }
        }
    }

    moves
}

/// Generate moves for a piece on a grid using its movement primitives.
#[allow(clippy::too_many_arguments)]
fn generate_grid_moves(
    session: &GameSession,
    zone_name: &str,
    cid: ComponentId,
    comp_def: &Component,
    col: u32,
    row: u32,
    adjacency: Option<Adjacency>,
    moves: &mut Vec<LegalMove>,
) {
    let zone = match session.runtime.zones.get(zone_name) {
        Some(z) => z,
        None => return,
    };
    let player = session
        .runtime
        .components
        .get(cid)
        .and_then(|c| c.owner.as_deref())
        .unwrap_or("");

    for mp in &comp_def.movement {
        if moves.len() >= MAX_LEGAL_MOVES {
            return;
        }
        match mp.primitive {
            PrimitiveType::Step => {
                let dirs = resolve_directions(mp, player, adjacency);
                let dist = mp.distance.unwrap_or(1);
                for (dx, dy) in &dirs {
                    let nx = col as i32 + dx * dist as i32;
                    let ny = row as i32 + dy * dist as i32;
                    if nx < 0 || ny < 0 || !zone.grid_cell_valid(nx as u32, ny as u32) {
                        continue;
                    }
                    if check_cell_condition(zone, session, nx as u32, ny as u32, mp, player) {
                        moves.push(LegalMove {
                            component_id: cid,
                            action: make_move_action(zone_name, col, row, nx as u32, ny as u32),
                        });
                    }
                }
            }
            PrimitiveType::Slide => {
                let dirs = resolve_directions(mp, player, adjacency);
                for (dx, dy) in &dirs {
                    let mut nx = col as i32 + dx;
                    let mut ny = row as i32 + dy;
                    while nx >= 0 && ny >= 0 && zone.grid_cell_valid(nx as u32, ny as u32)
                        && moves.len() < MAX_LEGAL_MOVES
                    {
                        let target = zone.grid_get(nx as u32, ny as u32);
                        if let Some(tid) = target {
                            if is_enemy(session, tid, player) {
                                moves.push(LegalMove {
                                    component_id: cid,
                                    action: make_move_action(
                                        zone_name, col, row, nx as u32, ny as u32,
                                    ),
                                });
                            }
                            break;
                        }
                        moves.push(LegalMove {
                            component_id: cid,
                            action: make_move_action(zone_name, col, row, nx as u32, ny as u32),
                        });
                        nx += dx;
                        ny += dy;
                    }
                }
            }
            PrimitiveType::Leap => {
                if let (Some(dx), Some(dy)) = (mp.dx, mp.dy) {
                    for (sx, sy) in &[(1, 1), (1, -1), (-1, 1), (-1, -1)] {
                        for (ldx, ldy) in &[(dx, dy), (dy, dx)] {
                            let nx = col as i32 + ldx * sx;
                            let ny = row as i32 + ldy * sy;
                            if nx < 0 || ny < 0 || !zone.grid_cell_valid(nx as u32, ny as u32) {
                                continue;
                            }
                            let target = zone.grid_get(nx as u32, ny as u32);
                            match target {
                                None => {
                                    moves.push(LegalMove {
                                        component_id: cid,
                                        action: make_move_action(
                                            zone_name, col, row, nx as u32, ny as u32,
                                        ),
                                    });
                                }
                                Some(tid) if is_enemy(session, tid, player) => {
                                    moves.push(LegalMove {
                                        component_id: cid,
                                        action: make_move_action(
                                            zone_name, col, row, nx as u32, ny as u32,
                                        ),
                                    });
                                }
                                _ => {}
                            }
                        }
                    }
                }
            }
            PrimitiveType::Hop => {
                let dirs = resolve_directions(mp, player, adjacency);
                for (dx, dy) in &dirs {
                    let mid_x = col as i32 + dx;
                    let mid_y = row as i32 + dy;
                    let land_x = col as i32 + dx * 2;
                    let land_y = row as i32 + dy * 2;
                    if mid_x < 0
                        || mid_y < 0
                        || land_x < 0
                        || land_y < 0
                        || !zone.grid_cell_valid(mid_x as u32, mid_y as u32)
                        || !zone.grid_cell_valid(land_x as u32, land_y as u32)
                    {
                        continue;
                    }
                    // Middle cell must be occupied
                    if zone.grid_get(mid_x as u32, mid_y as u32).is_none() {
                        continue;
                    }
                    // Landing cell must be empty
                    if zone.grid_get(land_x as u32, land_y as u32).is_none() {
                        moves.push(LegalMove {
                            component_id: cid,
                            action: make_move_action(
                                zone_name, col, row, land_x as u32, land_y as u32,
                            ),
                        });
                    }
                }
            }
            PrimitiveType::Place => {
                // Placement from supply onto empty cells — handled elsewhere
            }
            PrimitiveType::Flip => {
                let comp_name = session
                    .runtime
                    .components
                    .get(cid)
                    .map(|c| c.string_id.clone())
                    .unwrap_or_default();
                moves.push(LegalMove {
                    component_id: cid,
                    action: Action {
                        action_type: ActionType::Flip,
                        component_id: Some(comp_name),
                        ..default_action()
                    },
                });
            }
            PrimitiveType::Remove => {
                let comp_name = session
                    .runtime
                    .components
                    .get(cid)
                    .map(|c| c.string_id.clone())
                    .unwrap_or_default();
                moves.push(LegalMove {
                    component_id: cid,
                    action: Action {
                        action_type: ActionType::Remove,
                        component_id: Some(comp_name),
                        ..default_action()
                    },
                });
            }
            PrimitiveType::Swap => {
                let dirs = resolve_directions(mp, player, adjacency);
                let dist = mp.distance.unwrap_or(1);
                for (dx, dy) in &dirs {
                    let nx = col as i32 + dx * dist as i32;
                    let ny = row as i32 + dy * dist as i32;
                    if nx < 0 || ny < 0 || !zone.grid_cell_valid(nx as u32, ny as u32) {
                        continue;
                    }
                    if let Some(target_id) = zone.grid_get(nx as u32, ny as u32) {
                        let comp_name = session
                            .runtime
                            .components
                            .get(cid)
                            .map(|c| c.string_id.clone())
                            .unwrap_or_default();
                        let target_name = session
                            .runtime
                            .components
                            .get(target_id)
                            .map(|c| c.string_id.clone())
                            .unwrap_or_default();
                        moves.push(LegalMove {
                            component_id: cid,
                            action: Action {
                                action_type: ActionType::Swap,
                                component_id: Some(comp_name),
                                swap_with: Some(target_name),
                                ..default_action()
                            },
                        });
                    }
                }
            }
            _ => {
                // Remaining primitives (draw, move_to, castle, promote)
                // require cross-zone or multi-piece coordination
            }
        }
    }
}

/// Generate placement moves (e.g. Go stone placement, tic-tac-toe marks).
fn generate_placement_moves(
    _session: &GameSession,
    _player: &str,
    _zone_name: &str,
    _zone_def: &crate::definition::Zone,
    _zone: &RuntimeZone,
    _moves: &mut Vec<LegalMove>,
) {
    // Placement from supply is game-specific; will be expanded
}

/// Generate moves for playing from a hand zone.
fn generate_hand_plays(
    _session: &GameSession,
    _player: &str,
    _zone_name: &str,
    _zone: &RuntimeZone,
    _moves: &mut Vec<LegalMove>,
) {
    // Hand plays are game-specific; will be expanded
}

// --- Helpers ---

/// Resolve direction names to (dx, dy) vectors, respecting zone adjacency.
fn resolve_directions(
    mp: &MovementPrimitive,
    _player: &str,
    adjacency: Option<Adjacency>,
) -> Vec<(i32, i32)> {
    use crate::definition::Direction;

    let dir = match &mp.direction {
        Some(d) => d,
        None => return adjacent_directions(adjacency),
    };

    match dir {
        Direction::Single(name) => direction_vectors(name, adjacency),
        Direction::Multiple(names) => {
            names.iter().flat_map(|n| direction_vectors(n, adjacency)).collect()
        }
        Direction::Custom(_) => adjacent_directions(adjacency),
    }
}

fn direction_vectors(name: &DirectionName, adjacency: Option<Adjacency>) -> Vec<(i32, i32)> {
    match name {
        DirectionName::Orthogonal => vec![(1, 0), (-1, 0), (0, 1), (0, -1)],
        DirectionName::Diagonal => {
            if adjacency == Some(Adjacency::Hex6) {
                // Hex grids have no separate diagonal; return the non-orthogonal hex neighbors
                vec![(-1, 1), (1, -1)]
            } else {
                vec![(1, 1), (1, -1), (-1, 1), (-1, -1)]
            }
        }
        DirectionName::Adjacent => adjacent_directions(adjacency),
        DirectionName::Forward => vec![(0, 1)],
        DirectionName::ForwardDiagonal => vec![(1, 1), (-1, 1)],
        DirectionName::Backward => vec![(0, -1)],
        DirectionName::BackwardDiagonal => vec![(1, -1), (-1, -1)],
    }
}

fn hex_directions() -> Vec<(i32, i32)> {
    vec![(-1, 0), (1, 0), (0, -1), (0, 1), (-1, 1), (1, -1)]
}

fn adjacent_directions(adjacency: Option<Adjacency>) -> Vec<(i32, i32)> {
    match adjacency {
        Some(Adjacency::Hex6) => hex_directions(),
        Some(Adjacency::Orthogonal4) => vec![(1, 0), (-1, 0), (0, 1), (0, -1)],
        _ => {
            vec![
                (1, 0),
                (-1, 0),
                (0, 1),
                (0, -1),
                (1, 1),
                (1, -1),
                (-1, 1),
                (-1, -1),
            ]
        }
    }
}

fn is_enemy(session: &GameSession, target_id: ComponentId, player: &str) -> bool {
    session
        .runtime
        .components
        .get(target_id)
        .and_then(|c| c.owner.as_deref())
        .is_some_and(|owner| owner != player)
}

fn check_cell_condition(
    zone: &RuntimeZone,
    session: &GameSession,
    col: u32,
    row: u32,
    mp: &MovementPrimitive,
    player: &str,
) -> bool {
    let occupant = zone.grid_get(col, row);
    let condition = mp.condition.as_deref().unwrap_or("empty_or_enemy");

    match condition {
        "empty" => occupant.is_none(),
        "enemy" => occupant.is_some_and(|id| is_enemy(session, id, player)),
        "empty_or_enemy" => {
            occupant.is_none()
                || occupant.is_some_and(|id| is_enemy(session, id, player))
        }
        other => {
            let cell_empty = occupant.is_none();
            let cell_enemy = occupant.is_some_and(|id| is_enemy(session, id, player));
            // Try CEL evaluation for complex conditions
            if let Some(result) = crate::cel::try_eval_move_condition(cell_empty, cell_enemy, other)
            {
                return result;
            }
            // Legacy fallback: allow if empty or enemy
            cell_empty || cell_enemy
        }
    }
}

fn default_action() -> Action {
    Action {
        action_type: ActionType::Pass, // overridden by caller
        authority: None,
        component_id: None,
        component_type: None,
        from: None,
        to: None,
        zone: None,
        count: None,
        promote_to: None,
        orientation: None,
        rotation: None,
        amount: None,
        side: None,
        dice_count: None,
        dice_type: None,
        swap_with: None,
        declaration: None,
        commitment: None,
        custom_data: None,
    }
}

fn make_move_action(zone: &str, from_col: u32, from_row: u32, to_col: u32, to_row: u32) -> Action {
    Action {
        action_type: ActionType::MovePiece,
        authority: None,
        component_id: None,
        component_type: None,
        from: Some(Position::Structured {
            zone: Some(zone.to_string()),
            cell: Some(format!("{},{}", from_col, from_row)),
            index: None,
        }),
        to: Some(Position::Structured {
            zone: Some(zone.to_string()),
            cell: Some(format!("{},{}", to_col, to_row)),
            index: None,
        }),
        zone: None,
        count: None,
        promote_to: None,
        orientation: None,
        rotation: None,
        amount: None,
        side: None,
        dice_count: None,
        dice_type: None,
        swap_with: None,
        declaration: None,
        commitment: None,
        custom_data: None,
    }
}
